Add display, descriptionWidth, textWrap and isInvalid props to EuiExpression

andreadelrio opened the pull request:
### Summary

This PR adds the `display`, `descriptionWidth`, `textWrap` and `isInvalid` props to `EuiExpression`. The improvements introduced in this PR come, in part, as response to [this](https://github.com/elastic/kibana/issues/64788) Kibana ticket.

There are three new sections in the docs. One where `display` and `descriptionWidth` are shown and two others for `textWrap` and `isInvalid` respectively.

**New: Column display**
![indicies](https://user-images.githubusercontent.com/4016496/83467121-e161fb80-a42d-11ea-8cf4-e84bc7113047.gif)

**New: Invalid state for both clickable and read-only `Expression`s**
![image](https://user-images.githubusercontent.com/4016496/83466866-423d0400-a42d-11ea-8a3e-caa3f2550a83.png)

**New: Text truncation**
![image](https://user-images.githubusercontent.com/4016496/83466915-5d0f7880-a42d-11ea-91b5-c02c851924e1.png)


### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- [x] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
- [x] Props have proper **autodocs**
- [x] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)** examples
- [x] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**
~- [ ] Checked for **breaking changes** and labeled appropriately~
~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

mdefazio commented:
This looks good.  The demo example wasn't showing the truncate prop however. And do we need a `min-width`? Perhaps just enough so 2 letter operators still show?

cchaos commented on src/components/expression/expression.tsx:
```diff
@@ -66,6 +66,10 @@ export type EuiExpressionProps = CommonProps & {
    * Turns the component into a button and adds an editable style border at the bottom
    */
   onClick?: MouseEventHandler<HTMLButtonElement>;
+  /**
+   * Truncates text
+   */
+  truncate?: boolean;
```
I'm thinking we may want to make this more scalable by turning it into an [enum](https://github.com/elastic/eui/blob/master/wiki/component-design.md#enums-and-maps). Then it's more explicit to the consumer how the text will behave. Something like:

```suggestion
  /**
   * Truncates text
   */
  textWrap: 'break-word' | 'truncate';
```

Then the default would be `'break-word'`. Which also means moving the `@include euiTextBreakWord` mixin to a specific modifier class.

cchaos commented on src/components/expression/_expression.scss:
```diff
@@ -27,6 +27,11 @@
 
+.euiExpression-isTruncate {
+  @include euiTextTruncate;
+  vertical-align: bottom;
```
Is vertical-align something that should actually be applied to the default selector not a modifier?

cchaos commented on src-docs/src/views/expression/truncate.tsx:
```diff
@@ -0,0 +1,105 @@
+      <EuiPopover
+        button={
+          <EuiExpression
+            // truncate={true}
```
I'm guessing you didn't mean to keep this commented out

miukimiu commented on src-docs/src/views/expression/truncate.tsx:
```diff
@@ -0,0 +1,105 @@
+  value: string;
+}
+
+export default class extends Component<TruncateProps, TruncateState> {
```
Just jumping here because I noticed that this example is a class component. As described in #3150 new examples should avoid using class components and instead use function components.

andreadelrio replied:
@miukimiu Thanks for catching that! I turned it into a function component.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

andreadelrio commented:
> This looks good. The demo example wasn't showing the truncate prop however. And do we need a `min-width`? Perhaps just enough so 2 letter operators still show?

@mdefazio not sure what you mean by this, can you elaborate? do you mean a min-width for the `description` (left side) or for `EuiExpression` as a whole?

andreadelrio commented:
@cchaos wondering if it's wrong having to set a width on `EuiExpression` in order to have truncation work?

mdefazio commented:
> do you mean a min-width for the description (left side) or for EuiExpression as a whole?

I meant originally for the description (left-side), however maybe it should be as a whole—not sure which is right though tbh. 

Some thoughts as I play around with this some more:
1. Should we allow the expressions to wrap as the container shrinks (assuming the expression itself can still fit on one line)? 
2. If the expression can't fit on one line, then we truncate it.
3. Does it become tricky then if a 2nd expression is added dynamically based on the choice of the first? Do we want this to stay on the same line? Or simply allow it to drop to a new one?  I know in my original mock I showed it on the same line, but I understand this presents some issues. Perhaps in this scenario, both expressions are treated as one whole so we can truncate the only the end of it without the 2nd expression breaking to a new line. Not sure if this is possible though, or doesn't just create more trouble than good.
![image](https://user-images.githubusercontent.com/3756330/81935313-75bf0800-95be-11ea-8bd5-371968eae72f.png)

3. If these are true, then I'm guessing we would need the tooltip on every element, since we wouldn't know when it gets truncated?

andreadelrio commented:
> 1. Should we allow the expressions to wrap as the container shrinks (assuming the expression itself can still fit on one line)?

@mdefazio Why would the container shrink? Like on mobile? I’m thinking the consumer could specify a custom behaviour in their implementation (e.g. truncate but break-word and wrap in smaller screens). I don’t have a strong opinion about this though.

> 2. If the expression can't fit on one line, then we truncate it.
> 3. Does it become tricky then if a 2nd expression is added dynamically based on the choice of the first? Do we want this to stay on the same line? Or simply allow it to drop to a new one?  I know in my original mock I showed it on the same line, but I understand this presents some issues. Perhaps in this scenario, both expressions are treated as one whole so we can truncate the only the end of it without the 2nd expression breaking to a new line. Not sure if this is possible though, or doesn't just create more trouble than good.
>    ![image](https://user-images.githubusercontent.com/3756330/81935313-75bf0800-95be-11ea-8bd5-371968eae72f.png)
> 4. If these are true, then I'm guessing we would need the tooltip on every element, since we wouldn't know when it gets truncated?

For 2, 3 and 4

Here’s the way I’m looking at this: when using EuiExpression(s), are you ok with the layout changing visually? With it spanning multiple lines? If yes, don't truncate. That’s what we have here

![image](https://user-images.githubusercontent.com/4016496/81977109-6780e400-95de-11ea-91f7-1bba9b3eb97a.png)


If not, truncate and maybe add a tooltip. See the example below:
![image](https://user-images.githubusercontent.com/4016496/81977130-6c459800-95de-11ea-84d6-08afaf308406.png)

Pick “average()” of “name that still fits in one line”
![image](https://user-images.githubusercontent.com/4016496/81977139-6ea7f200-95de-11ea-99e9-f481f14a491a.png)

Pick “average()” of “long name that no longer fits in one line”￼
![image](https://user-images.githubusercontent.com/4016496/81977163-76679680-95de-11ea-8f24-5008ddeff1a6.png)

You don’t want that ^ to happen so you truncate:
![image](https://user-images.githubusercontent.com/4016496/81977175-79fb1d80-95de-11ea-954b-381666782c77.png)

In the case above (“WHEN” + “OF”), while we don’t know how long the content will be (dynamically inserted fields) we can know whether or not we want the user to see jumps in the layout (wrapping) or not. And for this same reason we’d know if something it’s going to get truncated and will need a tooltip. 

Let me know what you think. You mentioned truncating in this example (like in your mocks) could present some issues. What are those?

cchaos commented:
My guess is that all values and descriptions should wrap like normal text (where there is white space), but only if the **value** is longer than the width of the container will the value truncate instead of breaking mid-word.

<img width="734" alt="Screen Shot 2020-05-14 at 16 44 08 PM" src="https://user-images.githubusercontent.com/549577/81984310-9e68f100-9602-11ea-92d4-04d76c8f436a.png">

andreadelrio commented:
> My guess is that all values and descriptions should wrap like normal text (where there is white space), but only if the **value** is longer than the width of the container will the value truncate instead of breaking mid-word.
> 
> <img alt="Screen Shot 2020-05-14 at 16 44 08 PM" width="734" src="https://user-images.githubusercontent.com/549577/81984310-9e68f100-9602-11ea-92d4-04d76c8f436a.png">

If I'm understanding correctly, our truncation example would go from
![image](https://user-images.githubusercontent.com/4016496/81987638-e2ea9180-95ee-11ea-902b-f3f12b8d7e9d.png)
to
![image](https://user-images.githubusercontent.com/4016496/81987614-d5350c00-95ee-11ea-905b-0e6920af7e19.png)

I like the idea of only truncating the value. But what if we don't want the expression to go from 1 line to 2 lines?

cchaos commented:
>But what if we don't want the expression to go from 1 line to 2 lines?

I don't think there's anyway around that, but if they wanted they can just add a utility class of `.eui-textNoWrap`

snide commented:
I'd recommend a small design session around this one (maybe tomorrow if we all have time). The problems we're seeing in alerting is because there are very regularly large strings in smaller containers. I'm not certain truncation solves the problems here, and likely it's the actual styling of expression at the moment sort of falls apart in these scenarios. The entire component might need a restyle (or at least have an alternative style to be more readable for alerting's complex scenarios). I think using the underline alone is not good enough for the "button" when we know that it will often wrap to multiple lines.

Very specifically for the documentation example I would recreate the problem you're seeing in alerting. Showing one item truncated doesn't show the full breadth of the problem and how it can cause problems.

andreadelrio commented:
@cchaos @snide We've gone back and forth between a few ideas and wanted to share this one with you. Hopefully with your feedback we can move the design forward.

**Screenshot**

![image](https://user-images.githubusercontent.com/4016496/82962750-efb79f80-9f75-11ea-9225-0f894e692474.png)

**Gif**

![expression_again](https://user-images.githubusercontent.com/4016496/82962671-b2eba880-9f75-11ea-85d9-b9d92068239c.gif)

**Extras**

And we also explored adding a small triangle on the right (this could be either the default view or the hover/focus view)
![image](https://user-images.githubusercontent.com/4016496/82963128-e0852180-9f76-11ea-9a1a-291c758206ff.png)

cchaos commented:
I think the concept is very close and I like the background color indicating the separate expressions. In this particular case, however, I don't think the table-like alignments help with the reading of the full expression. 

Column layouts like this help to compare things in columns, however it makes it harder to read from left to right when there's so much space between the key/value pairs. 

![image](https://user-images.githubusercontent.com/549577/83057191-965a7980-a024-11ea-8799-1b3ab7957f64.png)

Instead, I would think you'd want the **values** to be in column layout so it's easy to compare the list, but keep the key/values close together similar to this:

<img width="865" alt="Screen Shot 2020-05-27 at 14 16 38 PM" src="https://user-images.githubusercontent.com/549577/83057265-b2f6b180-a024-11ea-8af2-757511cb69be.png">

---

For the focus state, I think our general focus ring would be enough to indicate focus state. The corner triangle starts to resemble the resize corner of a textarea

<img width="716" alt="Image 2020-05-27 at 2 19 16 PM" src="https://user-images.githubusercontent.com/549577/83057552-0ff26780-a025-11ea-98ed-568994c21e89.png">

mdefazio commented:
In regards to the table alignment, we had discussed previously about right-aligning the keys. I agree this might be less readable, but it would help avoid the jumping of the position of the value based on the length of the key.

Showing the read-only, error and valid states here as well.  Does the read-only still look too clickable?

![image](https://user-images.githubusercontent.com/3756330/83146111-cd339c80-a0c3-11ea-9a83-a814aeb8dd62.png)

snide commented:
@mdefazio I like that layout a lot better. I also like the spacing / styling a bunch.

andreadelrio commented:
> In regards to the table alignment, we had discussed previously about right-aligning the keys. I agree this might be less readable, but it would help avoid the jumping of the position of the value based on the length of the key.
> 
> Showing the read-only, error and valid states here as well. Does the read-only still look too clickable?
> 
> ![image](https://user-images.githubusercontent.com/3756330/83146111-cd339c80-a0c3-11ea-9a83-a814aeb8dd62.png)

I also like how it looks with the description aligned to the right. Two questions:

1) The read-only style looks too similar to the clickable one. For the background we're using the color that we usually use for disabled states (`$euiColorLightestShade`). Should we make it so the text is grayed out? @mdefazio

![image](https://user-images.githubusercontent.com/4016496/83169774-5f3da400-a0c8-11ea-97af-2594320d477a.png)


2) For the states (valid, error, etc.), that'd be something new in `EuiExpression`. To implement that, would it be like an append where the consumer can pass any node they want? Or do we define states and their styles/icons? @cchaos

mdefazio commented:
@andreadelrio , I went back and forth on the read-only state and agree with your comments.  Maybe the gray is too 'disabled' looking to me—maybe simply dark text for both?  I look at the current Expression implementation and it's a little more of a subtle difference between the two. So I wonder how far we need to push it?

![image](https://user-images.githubusercontent.com/3756330/83180590-a8561e00-a0f1-11ea-948c-0a0277588730.png)

One thing that's really nice about the Expression doc example is that it is filled out with defaults, so there's no empty states—guess that gets tricky with indices though?. The currently Alerting one does have error states, so maybe we just need that and not the 'valid' state. I'm not tied to the icons...just a thought for those with color blindness (the dark green and the red would appear quite similar, right?).

andreadelrio commented:
> @andreadelrio , I went back and forth on the read-only state and agree with your comments. Maybe the gray is too 'disabled' looking to me—maybe simply dark text for both? I look at the current Expression implementation and it's a little more of a subtle difference between the two. So I wonder how far we need to push it?
> 
> ![image](https://user-images.githubusercontent.com/3756330/83180590-a8561e00-a0f1-11ea-948c-0a0277588730.png)
> 
> One thing that's really nice about the Expression doc example is that it is filled out with defaults, so there's no empty states—guess that gets tricky with indices though?. The currently Alerting one does have error states, so maybe we just need that and not the 'valid' state. I'm not tied to the icons...just a thought for those with color blindness (the dark green and the red would appear quite similar, right?).

Ahh! Just checked what Alerting is doing. They're using `danger` for `color` whenever there's an error. 
Here's how using the `color` prop would look for this case (ignore the weird popover size). @mdefazio what do you think?

![errors](https://user-images.githubusercontent.com/4016496/83184427-4d66fb80-a0de-11ea-9a5b-e4546602f2d3.gif)

cchaos commented:
## Interaction states

While I think the original clickable vs readonly version (dashed underline) was actually more apparent for discerning between the two states, it didn't work well if the expression wrapped to more than a single line.

I'd be very wary of changing the readonly version to look disabled, I think these are two very different concepts. For instance, consumers may use the readonly version because they're just trying to explain a non-editable concept. I'd find a better way to signify this thing is clickable but this isn't.

## Validations states

I think the coloring goes a long way, but agree that the addition of the icons is really helpful especially for those with red/green color blindness. I'd opt for just a single icon to represent invalid, and not worry about the valid state.

## Column layout

I could see this being on specific option of EuiExpression. What you will have to do when implementing it is allow the consumer to control how wide the key side can be and then let the value take up the rest of the space.

andreadelrio commented:
I still have a couple of things to polish and fix but I've pushed my changes so you can all see the current status. I've added three props to the component: `display` (which can be the default `inline` or the new one `columns`), `descriptionWidth` and `isInvalid`.

![image](https://user-images.githubusercontent.com/4016496/83222240-ac536180-a12c-11ea-95df-ce4a60c2a6df.png)

![demodemodemodemo](https://user-images.githubusercontent.com/4016496/83222964-8dee6580-a12e-11ea-8d13-02d59f4f7743.gif)

I added a border to the clickable expressions to make a distinction with the read-only ones. This is just an exploration and we can explore more routes. @mdefazio

mdefazio commented:
This is really coming together! 
I think the border adds a bit too much visual noise. Maybe the description is simply `euiTextColor` for read-only expressions?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

andreadelrio commented:
> There also still seems to be a bit of a truncation issue. It might still be nice for consumers to be able to adjust this with a simple prop. Similar to how [EuiCodeEditor handles `whiteSpace`](https://github.com/elastic/eui/blob/29174cb5f9f663f91c014ca6daed9b3fbdf443a3/src/components/code/_code_block.tsx#L84-L89) you do for `wordWrap: 'truncate' | 'breakWord'.
> 
> <img alt="Screen Shot 2020-05-29 at 17 20 10 PM" width="323" src="https://user-images.githubusercontent.com/549577/83306945-8b911780-a1d1-11ea-9f90-23de78d22743.png">

@cchaos do you think that truncation should be for both the `description` and the `value`? or just the `value`? I would think that just for the `value`.

![image](https://user-images.githubusercontent.com/4016496/83376900-c4272180-a388-11ea-8f90-0f2e1e4a9220.png)

And another doubt, what do you think the truncate behaviour should be on the second Expression here ^? Truncate so that it only takes up one line? Like this
![image](https://user-images.githubusercontent.com/4016496/83377022-26802200-a389-11ea-95db-904f9dc61924.png)

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

cchaos commented:
I would assume you'd truncate both if they got so wide that they extended the full width of the container. If this isn't exactly how the consumer wants the behavior, they can add custom wrappers to their own implementation.

![Image 2020-06-01 at 11 32 51 AM](https://user-images.githubusercontent.com/549577/83425922-6af8d580-a3fc-11ea-90bf-95fdb99b9acf.png)

But I think it's more obvious what the component will do unless you specify a different wrap/truncation prop for each part (but we shouldn't need to go down that route).

andreadelrio commented:
@cchaos I added the `textWrap` prop and added a new example for it.

![image](https://user-images.githubusercontent.com/4016496/83440217-62eb6680-a3f9-11ea-800a-2ec338d63bc4.png)

andreadelrio commented:
@cchaos @mdefazio this should be ready for review. Can't re-request so pinging you this way. Thanks

cchaos commented on src-docs/src/views/expression/expression_example.js:
```diff
@@ -114,5 +141,81 @@ export const ExpressionExample = {
+      text: (
+        <div>
+          <p>
+            There might be cases where displaying a number of{' '}
```
```suggestion
            There might be cases where displaying multiple{' '}
```

cchaos commented on src-docs/src/views/expression/expression_example.js:
```diff
@@ -114,5 +141,81 @@ export const ExpressionExample = {
+            There might be cases where displaying a number of{' '}
+            <strong>EuiExpression</strong>s in a paragraph is not ideal. For
+            example, when both the <EuiCode>description</EuiCode> and the{' '}
+            <EuiCode>value</EuiCode> are variable or when the values they can
+            take can be quite long. To use a columns layout instead, pass{' '}
```
```suggestion
            <EuiCode>value</EuiCode> are variable or when their text is 
            quite long. To use a column display instead, pass{' '}
```

cchaos commented on src-docs/src/views/expression/expression_example.js:
```diff
@@ -114,5 +141,81 @@ export const ExpressionExample = {
+            example, when both the <EuiCode>description</EuiCode> and the{' '}
+            <EuiCode>value</EuiCode> are variable or when the values they can
+            take can be quite long. To use a columns layout instead, pass{' '}
+            <EuiCode>columns</EuiCode> to the <EuiCode>display</EuiCode> prop.
```
I find it easier for the consumer to read the right way to pass the prop when it's displayed as `prop="value"`.

```suggestion
            <EuiCode language="ts">{'display="columns"'}</EuiCode>.
```

cchaos commented on src-docs/src/views/expression/expression_example.js:
```diff
@@ -114,5 +141,81 @@ export const ExpressionExample = {
+            <EuiCode>columns</EuiCode> to the <EuiCode>display</EuiCode> prop.
+          </p>
+          <p>
+            The <EuiCode>description</EuiCode> column is aligned to the right.
```
```suggestion
            In column display, each expression is its own line and the <EuiCode>description</EuiCode> column is aligned to the right.
```

cchaos commented on src-docs/src/views/expression/expression_example.js:
```diff
@@ -114,5 +141,81 @@ export const ExpressionExample = {
+          </p>
+          <p>
+            The <EuiCode>description</EuiCode> column is aligned to the right.
+            The default width for it is 20%. You can customize this width via
```
```suggestion
            The default width for the description is 20%, but you can customize this with
```

cchaos commented on src-docs/src/views/expression/expression_example.js:
```diff
@@ -114,5 +141,81 @@ export const ExpressionExample = {
+            The default width for it is 20%. You can customize this width via
+            the <EuiCode>descriptionWidth</EuiCode> prop. When displaying a
+            group of <strong>EuiExpression</strong>s, make sure to set the same
+            width for all of them.
```
```suggestion
            width for all descriptions.
```

cchaos commented on src-docs/src/views/expression/expression_example.js:
```diff
@@ -114,5 +141,81 @@ export const ExpressionExample = {
+      text: (
+        <p>
+          Set <EuiCode>isInvalid</EuiCode> to true to display{' '}
+          <strong>EuiExpression</strong>&apos;s error state.
```
Add something like:

> The invalid state will override the `color` property with `danger`.

cchaos commented on src-docs/src/views/expression/stringing.tsx:
```diff
@@ -24,6 +24,9 @@ export default () => (
       description="group by"
       value="right.kytccountynmbr"
       onClick={() => {}}
+      color="accent"
+      isInvalid
     />
+    <EuiExpression description="sort by" value="count" isInvalid />
```
I think you can revert this change now that you have a specific section for showing invalid.

cchaos commented on src-docs/src/views/expression/truncate.js:
```diff
@@ -0,0 +1,26 @@
+    />
+    <EuiSpacer />
+    <EuiExpression
+      description="some very long description"
```
Did you mean to use the `description` variable here?

cchaos commented on src/components/expression/expression.test.tsx:
```diff
@@ -83,6 +83,60 @@ describe('EuiExpression', () => {
     });
 
+    describe('display', () => {
+      test('changes the layout', () => {
```
```suggestion
      test('can be columns', () => {
```

cchaos commented on src/components/expression/expression.test.tsx:
```diff
@@ -83,6 +83,60 @@ describe('EuiExpression', () => {
+    });
+
+    describe('isInvalid', () => {
+      test('true renders error state', () => {
```
```suggestion
      test('renders error state', () => {
```

cchaos commented on src/components/expression/expression.test.tsx:
```diff
@@ -83,6 +83,60 @@ describe('EuiExpression', () => {
+    });
+
+    describe('textWrap', () => {
+      test('truncate truncates text', () => {
```
```suggestion
      test('can truncate text', () => {
```

cchaos commented on src/components/expression/expression.tsx:
```diff
@@ -35,10 +36,20 @@ const colorToClassNameMap = {
 };
 
+const textWrapToClassNameMap = {
+  'break-word': 'euiExpression-breakWord',
```
I would leave off creating a class for `break-word` because you aren't actually using that class to add the appropriate styles. Instead, it's the way the default renders. Therefore, you only need to supply an additive class for the `truncate` value.

```suggestion
  'break-word': null,
```

cchaos commented on src/components/expression/expression.tsx:
```diff
@@ -35,10 +36,20 @@ const colorToClassNameMap = {
 
+const textWrapToClassNameMap = {
+  'break-word': 'euiExpression-breakWord',
+  truncate: 'euiExpression-truncate',
```
Reminder, in BEM, double dashes `--` are used for modifier classes where a single dash `-` is for state.
```suggestion
  truncate: 'euiExpression--truncate',
```

cchaos commented on src/components/expression/expression.tsx:
```diff
@@ -35,10 +36,20 @@ const colorToClassNameMap = {
 export type ExpressionColor = keyof typeof colorToClassNameMap;
 
+const displayToClassNameMap = {
+  inline: 'euiExpression-inline',
+  columns: 'euiExpression-columns',
```
Same comments as for the `textWrap` prop apply here.

```suggestion
  inline: null,
  columns: 'euiExpression--columns',
```

cchaos commented on src/components/expression/expression.tsx:
```diff
@@ -66,6 +77,25 @@ export type EuiExpressionProps = CommonProps & {
+   */
+  descriptionWidth?: number | string;
+  /**
+   * Sets how to handle text truncation.
```
```suggestion
   * Sets how to handle the wrapping of long text
```

cchaos commented on src/components/expression/_expression.scss:
```diff
@@ -21,6 +21,31 @@
+      vertical-align: bottom;
+    }
+
+    .euiExpression__value * {
+      // Covers the case where the value is made up of nodes
+      @include euiTextTruncate;
+      display: inline-block;
+    }
```
This could have unintended consequences, especially since it's applied to **all** children within `value`. I would remove this style block and instead, make a remark in the example description that it will only work properly if the prop types of `description` and `value` are strings and let them know, that otherwise, they need to use the `.eui-textTruncate` utility class on all their sub-children.

For instance, your indice example ends up pushing all the `p` tags into a single line which you don't want (as a consumer) since you added these `p` tags explicitly to separate each value onto its own line. 

<img width="538" alt="Screen Shot 2020-06-01 at 15 43 13 PM" src="https://user-images.githubusercontent.com/549577/83447569-a73d2d80-a41e-11ea-92a1-d2ba3f9e6e22.png">


I'd update that example, or add one to the truncate example, using the `.eui-textTruncate` utility class on the `p` tags.

andreadelrio commented on src-docs/src/views/expression/expression_example.js:
```diff
@@ -43,6 +43,33 @@ const stringingSnippet = `<div>
+const columnsSource = require('!!raw-loader!./columns');
+const columnsHtml = renderToHtml(Columns);
+const columnsSnippet = `<EuiExpression
+  description="description"
```
You're right. I updated all the snippets to use a JS variable for the `description`.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/

cchaos commented on src-docs/src/views/expression/truncate.js:
```diff
@@ -0,0 +1,48 @@
+const value = 'and a very long string as value';
+const description = 'some very very long description';
+const nodes = (
+  <div>
+    <p className="eui-textTruncate">.kibana_task_manager</p>
+    <p className="eui-textTruncate">kibana_sample_data_ecommerce</p>
+  </div>
```
```suggestion
  <>
    <p className="eui-textTruncate">.kibana_task_manager</p>
    <p className="eui-textTruncate">kibana_sample_data_ecommerce</p>
  </>
```

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3467/
